FBDeviceControl/Commands/FBDeviceApplicationCommands.m

```diff
@@ -237,6 +237,26 @@ - (instancetype)initWithDevice:(FBDevice *)device
 }
 
 - (FBFuture<id<FBLaunchedProcess>> *)launchApplication:(FBApplicationLaunchConfiguration *)configuration
+{
+  if (configuration.launchMode == FBApplicationLaunchModeFailIfRunning) {
+    return [[self processIDWithBundleID:configuration.bundleID] onQueue:self.device.asyncQueue chain:^ (FBFuture<NSNumber *>* processIdQueryResult) {
+      if (processIdQueryResult.state == FBFutureStateDone) {
+        return [[FBDeviceControlError
+          describeFormat:@"Application %@ already running with pid %@", configuration.bundleID, processIdQueryResult.result]
+          failFuture];
+      } else if (processIdQueryResult.state == FBFutureStateFailed) {
+        return (FBFuture*)[self launchApplicationIgnoreCurrentState:configuration];
+      } else {
+        return (FBFuture*)processIdQueryResult;
+      }
```

Comment:
The final branch herer looks a little suspicious. If the app is running then we fail, but in all other conditions we should probably attempt to launch the app

Reply:
You're probably right but maybe we should still not launch it if the future state is cancelled?

Reply:
Actually you might be right and we should just propagate the cancel. @lawrencelomax I'm guessing cancelling will be so rare that either is fine